testnet/stacks-node/src/neon_node.rs

```diff
@@ -735,9 +735,14 @@ fn spawn_peer(
                         }
                     }
                     Err(e) => {
-                        error!("P2P: Failed to process network dispatch: {:?}", &e);
-                        if config.is_node_event_driven() {
-                            panic!();
+                        if let NetError::Transient(ref msg) = &e {
+                            error!("P2P: Failed to process network dispatch due to transient error: {}", &msg);
+                        }
```

Comment:
Are you sure that these transient errors won't break the event dispatcher's guarantees? The reason that this panics in `event_driven` mode is if an error would cause a given block or block's events to be sent multiple times to the same observer. That would break the event dispatcher's interface contract with observers.

Reply:
> if an error would cause a given block or block's events to be sent multiple times to the same observer

Wait, is this part of the API contract?  If so, this sounds incredibly brittle, especially since an event observer can easily avoid re-processing a block's events if it's already seen the block.  Moreover, this API contract can't be guaranteed in the first place, since the same Stacks block can be processed on multiple PoX forks under normal circumstances.  Namely, if there are two identical Stacks blocks in two different PoX forks that share a parent block, then both Stacks blocks can be processed in their respective PoX forks since the state root calculation only depends on the _parent_ being in the same PoX fork as the miner when the miner calculated the block header.  If the event observer can handle this case already -- i.e. by using the block's consensus hash or index hash to distinguish two otherwise-identical blocks -- then it can easily handle the case where a block's receipts get sent more than once.

My impression was that the API contract only requires that the node not _drop_ block receits, since the observer has no way to poll for them (the node doesn't store them nor replay them).

Reply:
This guarantee is provided for `(consensus_hash, block_hash)` pairs. I'm not sure if that guarantee is strictly necessary, but this panic would protect the event observer from "dropped" data as well -- but I think we should get some insight from @lgalabru as well -- looking at the git history for this line, I think this was added for Atlas attachments. Could an error from the network.run command cause an attachment event to be missed and never reprocessed?

Reply:
I've made it so that `network.run()` does not fail from `net::Error::Transient(..)` -- all the call sites where this error could be emitted simply absorb it and turn it into an `Ok(..)`.